Throw pre-instanced IOException on connection reset

normanmaurer opened the pull request:
Motivation:

In the native transport we should throw a pre-instanced IOException on connection reset while reading.

Modifications:

Correctly throw pre-instanced IOException when ECONNRESET is received

Result:

Less overhead on connection reset

normanmaurer commented:
Fixed #3438 ... @trustin @Scottmitch please review

Scottmitch commented on transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java:
```diff
@@ -94,23 +95,29 @@
+        CONNECTION_RESET_EXCEPTION_WRITE = newConnectionResetWriteException("write");
+        CONNECTION_RESET_EXCEPTION_WRITEV = newConnectionResetWriteException("writev");
+        CONNECTION_RESET_EXCEPTION_READ = newConnectionResetReadException("read");
+        CONNECTION_RESET_EXCEPTION_SENDFILE = newConnectionResetWriteException("sendfile");
```
Send vs. Write terminology.  Can we make consistent?

normanmaurer replied:
you mean with the method names ? The thing is that for write and send it is the same ;)

Scottmitch replied:
Yes a bit of a mess cause there are calls to `send*` and `write*` functions where these are used...but now we are tying the two together (send and write).  Not the end of the world but it would be nice to use terminology agnostic to `write` or `send` because of this.  I'm not sure if introducing new terminology would help or make things worse but for example `newConnectionResetOutboundException` or maybe just leave at `newConnectionResetException`?  WDYT?

trustin replied:
I guess it's fine as long as it maps correctly to the corresponding system call names. We might want to add more information to the operation name so that it is clear it's the syscall names. e.g. (`syscall:writev()`)

Scottmitch commented on transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java:
```diff
@@ -120,7 +127,7 @@ private static int ioResult(String method, int err, IOException resetCause) thro
             return 0;
         }
-        if (err == ERRNO_EPIPE_NEGATIVE) {
+        if (err == ERRNO_EPIPE_NEGATIVE || err == ERRNO_ECONNRESET_NEGATIVE) {
```
Does this addition resolve (or change) the TODO comment on line 136?

normanmaurer replied:
nope...

Scottmitch commented:
LGTM! Good work as usual.

trustin commented:
Looks good to me. Please consider the comment above about exception message issues.

normanmaurer commented:
Addressed comments of @trustin and @Scottmitch ...

Cherry-picked into 4.0 (29a5bc9fb61e5d2c458dede927f0ee2482f037f5), 4.1 (91538d9b53976e4530123934458aec5c15ef56fe) and master (c3f7a71a735f0df6188404aa87447735d11673c9)
